pkg/asset/quota/openstack/openstack.go

```diff
@@ -11,44 +14,90 @@ import (
 // These constraints can be used to check if there is enough quota for creating a cluster
 // for the install config.
 func Constraints(ci *validation.CloudInfo, controlPlanes []machineapi.Machine, computes []machineapi.MachineSet) []quota.Constraint {
-	ctrplConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(controlPlanes))
-	for i, m := range controlPlanes {
-		ctrplConfigs[i] = m.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	var constraints []quota.Constraint
+
+	for i := 0; i < len(controlPlanes); i++ {
+		constraints = append(constraints, machineConstraints(ci, &controlPlanes[i])...)
+	}
+	constraints = append(constraints, instanceConstraint(int64(len(controlPlanes))))
+
+	for i := 0; i < len(computes); i++ {
+		constraints = append(constraints, machineSetConstraints(ci, &computes[i])...)
+	}
+	constraints = append(constraints, instanceConstraint(int64(len(computes))))
+
+	return aggregate(constraints)
+}
+
+func getOpenstackProviderSpec(spec *machineapi.ProviderSpec) *openstackprovider.OpenstackProviderSpec {
+	// An invalid ProviderSpec would presumably be a programming error here.
+	if spec.Value == nil {
+		logrus.Warnf("Empty ProviderSpec")
```

Comment:
A warning makes sense here, now that the platform-level flavor property is deprecated :+1:

Reply:
Let me unroll this comment.

IIUC, there are two valid cases:
* `spec.Value` is either `nil`; OR
* `spec.Value.Object` is of type `*openstackprovider.OpenstackProviderSpec`.

As a consequence:
* if `spec.Value` is nil, we accept (and warn) because the properties could be defined in the deprecated properties of the `platform` section;
* if `spec.Value.Object` is not OpenStack, we panic.

In other words: I agree with the code that I commented here (`logrus.Warnf`).

Is that any clearer, and agreeable?